FEAT : 닉네임 중복확인과 캐릭터 수정 API 구현

dy-p-code opened the pull request:
---
name: Pull Request
about: FEAT : 닉네임 중복확인과 캐릭터 수정 API 구현
title: FEAT : 닉네임 중복확인과 캐릭터 수정 API 구현
labels: backend, enhancement
assignees: 편도영

---

**작업한 내용**
닉네임 중복확인 API

**REVIEW POINT**
39eca44 커밋 내용 => FEAT : 닉네임 중복확인 API 수정 / 닉네임과 캐릭터 수정 API 구현

dy-p-code committed:
FEAT : 랭킹 찾기, 못찾겠다 꾀꼬리

dy-p-code committed:
FEAT : 유저 오늘 스코어 / 유저 오늘 랭킹 / 유저 토탈 스코어 API 구현

jn33-dev committed:
FIX : 이중 로그인시, 기존에 로그인했던 소켓 disconnect 처리 #115

- 이중 로그인시 userId로 socketIdMap에서 기존 로그인한 socket 정보를 찾아와 해당 소켓을 disconnect 처리
- 기존 socket의 클라로 error 메세지 송출
- 게임 종료 후 updateRoomInfoToRoom announce를 위해 event.user.info에 null 값 허용하도록 수정
Changes to be committed:
	modified:   src/games/games.gateway.ts
	modified:   src/games/players.service.ts
	modified:   src/games/util/event.user.info.constructor.ts

RESOLVE : #115

jn33-dev committed:
FIX : socket 이중 로그인 error 메세지 status code 변경 401-> 409 #115

Changes to be committed:
	modified:   src/games/games.gateway.ts

dy-p-code committed:
FEAT : 수정 완료료

dy-p-code committed:
BUILD : 주석 제거, 빌드 후 재커밋

dy-p-code committed:
FEAT : 닉네임 중복확인 API

dy-p-code committed:
FEAT :

dy-p-code committed:
CHORE : 주석 제거

jn33-dev commented on project/src/users/users.service.ts:
```diff
@@ -206,4 +206,33 @@ export class UsersService {
+            where: { nickname },
+        });
+
+        if (overlapCheck) {
```
위에서 **find** 함수로 찾으셨기 때문에, db에 동일 닉네임이 없더라도overlapCheck 변수에 []이 할당될겁니다.
if(overlapCheck) 조건 검사는 항상 true를 반환하게 된다는 얘기입니다.
console 찍어서 확인 한번 해주세요!

jn33-dev replied:
그리고 애초에 db에서 nickname이 unique 값으로 설정 되어 있기 때문에, find를 할 필요 없이findOne으로 변경하시면, 지금 올려주신 조건문을 그대로 사용할 수 있습니다.

jn33-dev commented on project/src/users/users.controller.ts:
```diff
@@ -69,11 +71,27 @@ export class UsersController {
+    }
+
+    // 닉네임/캐릭터 수정 API
+    @Get('/me/update')
```
@UseGuards(JwtAuthGuard) 붙여주세요

jn33-dev commented on project/src/users/users.controller.ts:
```diff
@@ -69,11 +71,27 @@ export class UsersController {
+
+    // 닉네임/캐릭터 수정 API
+    @Get('/me/update')
+    async userInfoChange(@Headers() headers, @Body() Body) {
```
@Body() **Body** -> @Body() **body** 소문자로 바꿔주세요.

jn33-dev replied:
Body가 class가 아니라 변수명인데, 여전히 대문자로 받아오시면...
**`userInfoChange(@Headers() headers, @Body() body)`**
이렇게 변경 부탁드립니다.

jn33-dev commented on project/src/users/users.controller.ts:
```diff
@@ -69,11 +71,27 @@ export class UsersController {
+    @Get('/me/update')
+    async userInfoChange(@Headers() headers, @Body() Body) {
+        const token = headers.authorization.replace('Bearer ', '');
+        await this.usersService.userInfoChange(token, Body);
```
여기서 Body가 {data: {nickname: stiring, profileImg: string}}으로 들어올텐데, 한 겹 까서 넣어야 되지 않을까요?
API 명세서 한번 확인해주세요~

jn33-dev replied:
프런트에서 해당 요청에 대한 req.body 값이 {nickname, profileImg} 로 들어오는지, {data: {nickname: stiring, profileImg: string}}로 들어오는지 확인 하신건가요?
혜연님과 반듯이 확인 부탁드립니다.

jn33-dev commented on project/src/users/users.service.ts:
```diff
@@ -206,4 +206,33 @@ export class UsersService {
+            throw new HttpException('해당하는 사용자를 찾을 수 없습니다.', 401);
+        } else
+            (userInfoChange) => {
+                const { nickname, profileImg } = Body;
```
controller 쪽에 Body 관련해서 커멘트 달아 두었습니다.
보시고 수정해주세요.

jn33-dev reviewed (changes requested):
도영님, 수정 사항 있어 커멘트 남겼습니다.
여전히 수정해야 할 것들은 여러군데 보이지만, 그래도 처음보다는 일취월장 하신 듯 합니다!
그리고, 해당 PR에서 두 건의 이슈를 해결하고 계시니, PR 제목도 작업하신 내용이 잘 보이도록 수정 부탁드립니다!

close #151 #150

dy-p-code committed:
FEAT : 닉네임/캐릭터 수정 API 수정

dy-p-code committed:
CHORE : 유저 토탈 스코어 조회를 위해 타입 변환

jn33-dev reviewed (changes requested):
도영님, 수정해야 할 것 1, 확인해야 할 것 1 커멘트 달았습니다.

dy-p-code committed:
CHORE : 수정 최종본이 올라가는지 확인

dy-p-code committed:
Merge branch 'dev' into feature/141-re

dy-p-code committed:
CHORE : 최종본으로 올라가는지 확인

dy-p-code committed:
CHORE : DB에서 수정이 됐다가 안됐다가 함

jn33-dev commented on project/src/users/users.service.ts:
```diff
@@ -230,8 +230,7 @@ export class UsersService {
                 const { nickname, profileImg } = body;
-                const bodyValue = Object.values({ nickname, profileImg });
-                this.usersRepository.save(bodyValue);
+                this.usersRepository.update({ nickname: nickname }, { profileImg: profileImg });
```
회원정보 수정이 됐다 안됐다 하는 이유는 두가지로 보입니다.
1. await이 빠져서 비동기 처리가 안됨
2. body 값으로 받는 nickname은 db에 없는 nickname이 들어오는 경우도 있는데, 지금 사용하신 update 방식이라면, nickname에 해당하는 모든 entity를 찾아 profileImg만 변경한다가 됩니다.
-> save 함수를 사용하시고, req.header에서 받은 token 값을 이용해 userId를 찾아서 진행해주세요.
save 함수로 db에 있는 데이터를 업데이트 하려면 반듯이 input 값에 PK 정보를 포함해야 됩니다.

dy-p-code committed:
FEAT : 진짜진짜 끝

dy-p-code committed:
Merge pull request #154 from gachimind/feature/141-re

FEAT : 유저 오늘 토탈 스코어 / 오늘 랭킹 / 게임 전체 스코어 API 구현

jn33-dev committed:
Merge branch 'dev' of https://github.com/gachimind/backend into feature/115

Conflicts:
	project/dist/tsconfig.build.tsbuildinfo

	modified:   src/users/users.controller.ts
	modified:   src/users/users.service.ts

jn33-dev committed:
FIX : 기존 소켓이 연결 끊긴 상태에서 새로운 소켓 정보로 로그인 시도 할때, 에러 핸들링 & socketIdMap 데이터 갱신 처리 #115

Changes to be committed:
	modified:   src/games/dto/socketId-map.request.dto.ts
	modified:   src/games/entities/socketIdMap.entity.ts
	modified:   src/games/games.gateway.ts
	modified:   src/games/players.service.ts
	modified:   src/users/users.service.ts

RESOLVE : #115

jn33-dev committed:
Merge pull request #155 from gachimind/feature/115

FIX : 이중 로그인시, 기존에 로그인했던 소켓 disconnect 처리 #115

jn33-dev committed:
FEAT : timezone 확인

On branch feature/115
Changes to be committed:
	modified:   src/games/util/today.date.constructor.ts

jn33-dev committed:
timzone test

jn33-dev committed:
timezone test2

jn33-dev committed:
FIX : socketIdMap PK 변경한 것에 기존 다른 로직과 충돌이 있어 revert #115

- log-in 이벤트에서 기존 로그인 정보가 있다면, socketIdMap에서 해당 정보 삭제 후 생성
Changes to be committed:
	modified:   src/games/dto/socketId-map.request.dto.ts
	modified:   src/games/entities/socketIdMap.entity.ts
	modified:   src/games/games.gateway.ts
	modified:   src/games/players.service.ts

RESOLVE : #115

jn33-dev committed:
Merge branch 'feature/150' of https://github.com/gachimind/backend into feature/150

Conflicts:
	# project/dist/src/users/users.controller.js.map
	# project/dist/src/users/users.service.js.map
	# project/dist/tsconfig.build.tsbuildinfo
	project/src/users/users.controller.ts
	project/src/users/users.service.ts

jn33-dev committed:
REFACTOR : [회원] createToken 메서드에서 변수명 등 변경 #150

Changes to be committed:
	new file:   src/users/dto/token.map.request.dto.ts
	modified:   src/users/entities/token-map.entity.ts
	modified:   src/users/users.controller.ts
	modified:   src/users/users.service.ts

jn33-dev committed:
Merge branch 'feature/115' of https://github.com/gachimind/backend into feature/150

Changes to be committed:
	modified:   src/games/dto/socketId-map.request.dto.ts
	modified:   src/games/entities/socketIdMap.entity.ts
	modified:   src/games/games.gateway.ts
	modified:   src/games/players.service.ts
	modified:   src/games/util/today.date.constructor.ts

jn33-dev committed:
FIX : socketIdMap softdelete를 위해 userInfo FK constraint 해제 #115

Changes to be committed:
	modified:   src/games/entities/socketIdMap.entity.ts

RESOLVE : #115

jn33-dev committed:
Merge branch 'feature/115' of https://github.com/gachimind/backend into feature/150

Changes to be committed:
	modified:   src/games/entities/socketIdMap.entity.ts

jn33-dev committed:
FEAT : 닉네임 중복확인과 캐릭터 수정 API 구현 #151

- 닉네임 검증 로직 추가 -> TODO 확인하여 class-validator 사용으로 변경하고, profileImg도 검증 로직 추가할 것
- 컨트롤러에서 Get method Post로 수정
- updateUser dto 수정
Changes to be committed:
	modified:   src/users/dto/update-user.dto.ts
	modified:   src/users/users.controller.ts
	modified:   src/users/users.service.ts

RESOLVE #151, #150

jn33-dev reviewed (approved):
- 닉네임 검증 로직 추가
- updateUser dto 수정
- 에러핸들링 추가
- Get -> Post 메서드로 수정

결과물 잘 나오는 것 확인했습니다.
